### DISCUSSION TITLE
Unable to launch Gaphor or run Pytests in a Docker container.

### DISCUSSION THREAD

**Discussion by jashangills** [gaphor/gaphor#3576] [General]
We are currently building a CI/CD pipeline using Docker and need to automate tests that depend on Gaphor. Specifically, our Pytests and some other scripts require the Gaphor GUI to be launched.

Since the tests are running inside Docker containers, we want to know if Gaphor supports any kind of headless GUI mode, allowing it to operate without launching the full graphical interface. This would enable us to run the tests in a headless environment without requiring access to a physical display.

If headless mode is not supported, are there any community-supported workarounds or best practices to run Gaphor within a Docker container?

### To Reproduce
Steps to reproduce the behaviour:
1.Build a custom Docker image using the official `ghcr.io/gaphor/gaphor-appimage`
2.Run the Docker container in interactive mode
3.Install the required dependencies using pip.
4.Attempt to launch Gaphor or run Pytests that depend on it.

### Expected behavior
Pytests should pass successfully by executing Gaphor-dependent tests in the Docker container.

### Screenshots
<img width="1324" alt="image" src="https://github.com/user-attachments/assets/07162f58-3a50-456e-b9e3-5a5580d609d6">

### OS
- [Y ] Linux (Docker env)
- [ ] macOS
- [ ] Windows

### Version
2.24.0

### Additional information
We are seeking guidance on whether Gaphor has official or community support for headless GUI operations. If this feature is unavailable, any recommended workarounds for automating tests in a Docker-based CI/CD pipeline would be greatly appreciated.

**Comment by amolenaar** [comment]
Hi @jashangills,

Thank you for your question. It's a good one.

Gaphor does work without a GUI. That's how we render [the jupyter notebooks](https://docs.gaphor.org/en/latest/scripting.html) in our documentation.

What you'd need is:

* A C compiler, to compile PyGObject
* Install `python3-devel`, `glib-devel`, `cairo-gobject-devel`,  `gobject-introspection-devel`, `pango`
* Install `graphviz` if you want to use auto-layouting
* `pip install pygobject` should work at this point
* `pip install gaphor` 
```

Build PyGObject from PyPI:

```Dockerfile
FROM fedora:40

RUN dnf install -y gcc python3-devel glib-devel cairo-gobject-devel  gobject-introspection-devel pango
RUN python3 -m venv venv
RUN venv/bin/pip install gaphor
```

Alternatively, use the system-provided PyGObject, and install everything at system level:

```Dockerfile
FROM fedora:40

RUN dnf install -y python3 python3-pip python3-gobject pango
RUN pip install gaphor
```

If you need to run GUI code, please have a look at our build scripts for some inspiration.

Note: I would not use the `ghcr.io/gaphor/gaphor-appimage` container, but rather use something more up to date, such as `fedora:41`.

**Comment by jashangills** [comment]
Thanks @amolenaar for your advice and prompt response.
Using Fedora and Installing the dependencies as mentioned in your build scripts, I was able to run the pytests and achieve a successful pipeline build.

**Reply by amolenaar** [reply]
@jashangills Thank you. I'm glad I could help.

**Comment by mikekidner** [comment]
@amolenaar / @danyeaw I'm trying to create a docker image for gaphor. I'm running into an issue with libgirepository-2.0-dev.
On mac-osx ubuntu 24.04 does not build so I have to use 22.04. - But libgirepository-2.0-dev does not install. 
What's the correct way out of this ?
Here's the dockerfile - 
```
# [Choice] Ubuntu version (use ubuntu-22.04 or ubuntu-18.04 on local arm64/Apple Silicon): ubuntu-22.04, ubuntu-20.04, ubuntu-18.04                                                                            
#ARG VARIANT=ubuntu-24.04                                                                                                                                                                                      
ARG VARIANT=ubuntu-22.04
FROM mcr.microsoft.com/vscode/devcontainers/base:0-${VARIANT} 
ENV DEBIAN_FRONTEND=noninteractive                                                                                                                                                                                                                                                                                                                                                                                         
# Install additional OS packages.                                                                                                                                                                              
RUN apt-get update -qq &&  apt-get install -y -qq --no-install-recommends \                                                                                                                                    
    libcairo2-dev \                                                                                                                                                                                            
    pkg-config \                                                                                                                                                                                               
    python3-dev \                                                                                                                                                                                              
    gir1.2-gtk-4.0 \                                                                                                                                                                                           
    libgtksourceview-5-dev \                                                                                                                                                                                   
    libadwaita-1-dev \                                                                                                                                                                                         
    #libgirepository-2.0-dev \                                                                                                                                                                                 
    && apt-get clean -y && rm -rf /var/lib/apt/lists/*                                                                                                                                                         
WORKDIR /app                                                                                                                                                                                                   
CMD ["/bin/bash"]
```
 I've commented out the libgirepository-2.0-dev line to get the image to build, then in the container running apt-get update and then install yields:
```
 apt-get install libgirepository-2.0-dev
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package libgirepository-2.0-dev
E: Couldn't find any package by glob 'libgirepository-2.0-dev'
E: Couldn't find any package by regex 'libgirepository-2.0-dev'
```

**Reply by danyeaw** [reply]
Hi @mikekidner - could you use a newer Fedora docker image instead? An old LTS version of Ubuntu is not going to work because the dependencies are just too old.

**Comment by amolenaar** [comment]
Hi @mikekidner, I think Ubuntu 22.04 is too old. libgirepository was introduced with GLib 2.80.

Not sure if Ubuntu 24 will work, though. We [use Fedora (41)](https://github.com/gaphor/gaphor/blob/main/.github/workflows/full-build.yml#L83) in the build pipeline, since Ubuntu 24 ships with an older version of GNOME than what Gaphor needs.

**Comment by mikekidner** [comment]
Thanks for the pointers, I got a image to build with Fedora41. 
After pip install gaphor from my local fork I encounter a segmentation fault when initialising a Session
```
from gaphor.application import Session
s = Session()
```

```
/app/gaphor/gaphor/plugins/console/consolewindow.py:5: PyGIWarning: Adw was imported without specifying a version first. Use gi.require_version('Adw', '1') before import to ensure that the right version gets loaded.
  from gi.repository import Adw, Gdk, Gtk
Settings schema not found and settings won’t be saved. Run `gaphor install-schemas`.
(process:674): Gtk-CRITICAL **: 22:50:18.150: gtk_icon_theme_get_for_display: assertion 'GDK_IS_DISPLAY (display)' failed
(process:674): Gtk-CRITICAL **: 22:50:18.151: gtk_icon_theme_add_resource_path: assertion 'GTK_IS_ICON_THEME (self)' failed
Segmentation fault
```
 This seems to occur when GTK builder is called. 
 Does my image have to do some display forwarding to start a session??

**Reply by mikekidner** [reply]
Solved. 
I can check for DISPLAY being defined and then not use Session directly , as I'm running without the UI. Then just following the jupyter notebook examples to create the element factory etc.
